File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.py
def solution(key, lock):
    def rotate_key():
        result = [[0] * m for _ in range(m)]
        for r in range(m):
            for c in range(m):
                result[c][m - 1 - r] = key[r][c]
        return result
    
    def check_key():
        for i in range(m - 1, m + n - 1):
            for j in range(m - 1, m + n - 1):
                if board[i][j] != 1:
                    return False
        return True
    
    n, m = len(lock), len(key)
    board = [[0] * (n + 2 * (m - 1)) for _ in range(n + 2 * (m - 1))]
    
    for i in range(m - 1, m + n - 1):
        for j in range(m - 1, m + n - 1):
            board[i][j] = lock[i - m + 1][j - m + 1]
            
    for i in range(n + m - 1):
        for j in range(n + m - 1):
            for _ in range(4):
                for x in range(m):
                    for y in range(m):
                        board[i + x][j + y] += key[x][y]
                
                if check_key():
                    return True
                
                for x in range(m):
                    for y in range(m):
                        board[i + x][j + y] -= key[x][y]
                
                key = rotate_key()

    return False

File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.java
class Solution {

    private static int n;
    private static int m;

    public boolean solution(int[][] key, int[][] lock) {
        n = lock.length;
        m = key.length;

        int size = n + 2 * (m - 1);
        int[][] board = new int[size][size];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[m - 1 + i][m - 1 + j] = lock[i][j];
            }
        }

        int[][][] keys = rotateAndSave(key);

        for (int i = 0; i < n + m - 1; i++) {
            for (int j = 0; j < n + m - 1; j++) {
                for (int r = 0; r < 4; r++) {
                    for (int x = 0; x < m; x++) {
                        for (int y = 0; y < m; y++) {
                            board[i + x][j + y] += keys[r][x][y];
                        }
                    }

                    if (isValid(board)) return true;

                    for (int x = 0; x < m; x++) {
                        for (int y = 0; y < m; y++) {
                            board[i + x][j + y] -= keys[r][x][y];
                        }
                    }
                }
            }
        }

        return false;
    }

    private static boolean isValid(int[][] board) {
        for (int i = m - 1; i < m - 1 + n; i++) {
            for (int j = m - 1; j < m - 1 + n; j++) {
                if (board[i][j] != 1) return false;
            }
        }
        
        return true;
    }

    private static int[][][] rotateAndSave(int[][] key) {
        int[][][] keys = new int[4][m][m];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < m; j++) {
                keys[0][i][j] = key[i][j];
            }
        }

        for (int k = 1; k < 4; k++) {
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < m; j++) {
                    keys[k][i][j] = keys[k - 1][m - j - 1][i];
                }
            }
        }

        return keys;
    }
}